implementations/elixir/ockam/ockam/lib/ockam/protocol/bare_extended.ex

```diff
@@ -126,17 +129,100 @@ defmodule Ockam.Bare.Variant do
   def enum_value(_tag), do: nil
 
   def encode_value(nil, nil), do: <<>>
-  def encode_value(value, {_tag, subschema}), do: :bare.encode(value, subschema)
+  def encode_value(value, {_tag, subschema}), do: Extended.encode(value, subschema)
+end
+
+defmodule Ockam.Bare.Tuple do
+  @moduledoc """
+  Support tuple types
+  Tuple types are defined as {:tuple, [schema(), ...]}
+  The values are encoded sequentially and are order sensitive.
+  """
+
+  @type schema :: :bare.spec()
+  @type extended_schema() :: schema() | {:tuple, [{atom(), schema()}]}
+
+  @spec encode(any(), extended_schema()) :: binary()
+  def encode([value | values], {:tuple, [type | types]} = _schema) do
+    encode(value, type) <> encode(values, {:tuple, types})
+  end
+
+  @spec encode(any(), extended_schema()) :: binary()
+  def encode([], {:tuple, []} = _schema) do
+    <<>>
+  end
+
+  @spec encode(any(), extended_schema()) :: binary()
+  def encode(something, {:tuple, []} = _schema) do
+    raise "Extra data in tuple: #{inspect(something)}"
+  end
+
+  @spec encode(any(), extended_schema()) :: binary()
+  def encode({}, {:tuple, [_type | _types]} = schema) do
+    raise "Missing data in tuple: #{inspect(schema)}"
+  end
+
+  @spec encode(any(), extended_schema()) :: binary()
+  def encode([value], [schema]) do
+    :bare.encode(value, schema)
+  end
+
+  @spec encode(any(), extended_schema()) :: binary()
+  def encode(value, schema) do
+    :bare.encode(value, schema)
+  end
+
+  @spec decode(binary(), extended_schema()) :: {:ok, any()} | {:error, any()}
+  def decode(<<>>, {:tuple, [{:optional, _}]} = _schema) do
+    {:ok, [:undefined], <<>>}
+  end
+
+  @spec decode(binary(), extended_schema()) :: {:ok, any()} | {:error, any()}
+  def decode(<<>>, {:tuple, []} = _schema) do
+    {:ok, [], <<>>}
+  end
+
+  @spec decode(binary(), extended_schema()) :: {:ok, any()} | {:error, any()}
+  def decode(data, {:tuple, [type | types]} = _schema) do
+    case decode(data, type) do
+      {:ok, value, rest} ->
+        with {:ok, values, ""} <- decode(rest, {:tuple, types}) do
+          {:ok, [value] ++ values, <<>>}
+        end
+
+      {:ok, value} ->
```

Comment:
something is fishy here.    If decode/2  returns `{:ok, value}` one would assume it would be because there is no more input.    In that case,  we shouldn't accept the thing,  _unless_  there are also no remaining non-optional fields on the schema (`types == []`).  Otherwise   things like `decode( encode([1,2], {:tuple, [:int, :int]}),   {:tuple, [:int, :int, :int, :int]})`  would return a 2 element tuple even when it should be 4 and none is optional.

*However,  that is not happening*.    I suspect because decode/2  _never_ returns `{:ok, value}` .   It returns `{:ok, value, ""}`.    That's what  `:bare.decode/2` returns.    

I'd suggest re-checking if that's the case,  if so fix the typespec and remove this clause.

Reply:
You were right! The clause was unnecessary, i've also updated the typespec to reflect the binary being returned.